### PR TITLE
Add kafka log parse utility classes/functions and write time sync kafka log parser

### PR DESCRIPTION
**Pull request by paulbourelly999**
# PR Details
## Description
Added python scripts to parse Kafka log data for **SPAT** and **Time Sync** messages into csv format.

## Related Jira Key
CDAR-535

## Motivation and Context
Data Analysis

## How Has This Been Tested?
Local Testing 

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/carma-streets/ParseTimeSyncToCSV.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+from csv import writer
+from pathlib import Path
+from enum import Enum
+import argparse
+import ParseKafkaLog
+from KafkaLogMessage import KafkaLogMessage, KafkaLogMessageType
+
+
+def parse_timesync_log(inputfile: Path, outputfile: Path):
```
Comment (adamlm) [inline src/carma-streets/ParseTimeSyncToCSV.py:9]: This function can be renamed to something more descriptive. Perhaps `extract_time_sync_msgs_to_csv()`.

File: src/carma-streets/ParseKafkaLog.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import json
+import re
+from pathlib import Path
+from enum import Enum
+from KafkaLogMessage import KafkaLogMessage, KafkaLogMessageType
+
+
+def parse_kafka_logs(inputfile: Path, message_type: KafkaLogMessageType)-> list:
```
[the remark this answers is not in the document]
Comment (paulbourelly999) [reply src/carma-streets/ParseKafkaLog.py:8]: I am not sure I agree that parse_kafka_logs is misleading. Maybe you explain to my what you interpret parse meaning and why you believe extract is more descriptive

File: src/carma-streets/ParseKafkaLog.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    """Parse Kafka Topic Logs into a list of KafkaLogMessages
+
+    Args:
+        inputfile (Path): Path to an inputfile
+        message_type (KafkaLogMessageType): Type of KafkaLogMessage to parse from input file
```
[the remark this answers is not in the document]
Comment (paulbourelly999) [reply src/carma-streets/ParseKafkaLog.py:13]: If they are not the correct message type we will encounter JSON parsing errors. This comment also does not claim it verifies message type. It just indicates that this is the message type in the input file. I can reword it to more clearly indicate that

File: src/carma-streets/ParseKafkaLog.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    Returns:
+        [KafkaLogsMessage]: list of KafkaLogMessages parsed from input file
+    """
+    if inputfile.is_file():
+        #Convert the text file into an array of lines
+        with open(inputfile, encoding="utf8", errors='ignore') as textFile:
```
[the remark this answers is not in the document]
Comment (paulbourelly999) [reply src/carma-streets/ParseKafkaLog.py:20]: Why does it matter for the quotations

File: src/carma-streets/ParseSPATToCSV.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
+from csv import writer
+from pathlib import Path
+from enum import Enum
+import argparse
+import ParseKafkaLog
+from KafkaLogMessage import KafkaLogMessage, KafkaLogMessageType
+
+
+def parseSpatToCsv(inputfile: Path, outputfile: Path):
```
[the remark this answers is not in the document]
Comment (paulbourelly999) [reply src/carma-streets/ParseSPATToCSV.py:9]: Could you justify why you prefer extract to parse?

File: src/carma-streets/ParseSPATToCSV.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    """Function to parse SPAT Kafka Topic log file and generate csv data of all time sync messages
+
+    Args:
+        inputfile (Path): Path to Kafka Topic log file
+        outputfile (Path): File name (excluding file extension) of desired csv file
+    """
+    spat_msgs = ParseKafkaLog.parse_kafka_logs(inputfile, KafkaLogMessageType.TimeSync)
```
[the remark this answers is not in the document]
Comment (paulbourelly999) [reply src/carma-streets/ParseSPATToCSV.py:16]: That is a bug.

**Comment by paulbourelly999** [comment]
> * functions and classes into separate files is unnecessary. They are all required for data parsing and are used only for this purpose (_i.e._, they are not library files).
> * Following the above point, the code can be combined into a single script, possibly called `extract_kafka_msgs_from_logs.sh`. It would look something like this:

Separating these improve modularity and reusability of parsing logic. My vision here is to have a python module that is responsible for parsing all carma streets kafka data regardless of use case. Then possibly have a dependent python module that is able to generate a lot of the common use case data analysis plots. Obscuring our implementation for data analysis presents very little value here to me, especially since this is no some script to be copied into 16 different common repos but more of a python package that will evolve with every data analysis

File: src/carma-streets/KafkaLogMessage.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+class KafkaLogMessage:
+    """Class used to store data for each Kafka Log Message
+    """
+    created: int
+    json_message: dict
+    msg_type: KafkaLogMessageType
+    def __init__(self, created, json_message, msg_type):
+        self.created = created
+        self.json_message = json_message
+        self.msg_type = msg_type
```
Comment (adamlm) [inline src/carma-streets/KafkaLogMessage.py:30]: This method is unnecessary for data classes.

File: src/carma-streets/ParseKafkaLog.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (adamlm) [reply src/carma-streets/ParseKafkaLog.py]: Consistency. It is a stylistic choice outside of nested quotes.

File: src/carma-streets/ParseKafkaLog.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (adamlm) [reply src/carma-streets/ParseKafkaLog.py:13]: > If they are not the correct message type we will encounter JSON parsing errors.

The function's name and signature implies that it will search the specified log file for messages of only type `message_type`, which is not the case. Instead, this function searches the log for _any_ valid JSON string that comes after `CreateTime` then creates a `KafkaLogMessage` instance of type `message_type`.

An uninformed user could do something like this:

```python
result = parse_kafka_logs(spat_log_path, KafakLogMessageType.BSM)
```
It is unclear what is actually happening or why this is problematic. The alternative is clearer:

```python
# Easier to identify a problem
result = extract_kakfa_msgs_as_type(spat_log_path, KafkaMessageType.BSM)

# Easier to understand what is happening
result = extract_kakfa_msgs_as_type(bsm_log_path, KafkaMessageType.BSM)
```

File: src/carma-streets/ParseKafkaLog.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (adamlm) [reply src/carma-streets/ParseKafkaLog.py]: All uses of the term _parsing_ that I can think of involve building a structured representation of data from a string (or similar type of source):

- compilers parse source code to build abstract syntax trees
- JSON/YAML/XML/CSV parsers build a structured representation of the data

All of these examples have one thing in common: they require the whole data source to be valid. Compilers reject code that has syntax errors. File format parsers fail to load when the files are malformed.

Our use here is different from that. We are extracting a subset of the data; we are extracting Kafka message data from log files. Therefore, I think the term `extract` is more appropriate.

File: src/carma-streets/ParseSPATToCSV.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (adamlm) [reply src/carma-streets/ParseSPATToCSV.py]: See above comment.

**Comment by adamlm** [comment]
> Separating these improve modularity and reusability of parsing logic.

Combining the functions and classes into a single Python module does not reduce modularity. They are all related to extracting Kafka messages from logs, so it would make more sense to combine them into a single module.

> My vision here is to have a python module that is responsible for parsing all carma streets kafka data regardless of use case. Then possibly have a dependent python module that is able to generate a lot of the common use case data analysis plots.

What you are describing here is similar to the analysis scripts for the Eclipse MOSAIC Real-Time Factor (see #19).

> Obscuring our implementation for data analysis presents very little value here to me, especially since this is no [sic] some script to be copied into 16 different common repos but more of a python package that will evolve with every data analysis

Why would we have to copy this to several repos?

File: src/carma-streets/ParseKafkaLog.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulbourelly999) [reply src/carma-streets/ParseKafkaLog.py]: We require the data source to be valid as well. Our parser actual includes JSON parsing and will throw exception for misformated JSON. 

I think definitions of parse I find online align well with what is happening here:
image
https://www.merriam-webster.com/dictionary/parser

File: src/carma-streets/ParseKafkaLog.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (paulbourelly999) [reply src/carma-streets/ParseKafkaLog.py]: Resolved

File: src/carma-streets/ParseKafkaLog.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (paulbourelly999) [reply src/carma-streets/ParseKafkaLog.py]: I added as type to the function name

**Comment by paulbourelly999** [comment]
@adamlm Please resolve any concerns that you feel have been addressed. I am unsure which of these comments are still relevant

File: src/carma-streets/ParseSPATToCSV.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (paulbourelly999) [reply src/carma-streets/ParseSPATToCSV.py]: addressed

File: src/carma-streets/ParseTimeSyncToCSV.py (hunk 1 above)
Comment (paulbourelly999) [reply src/carma-streets/ParseTimeSyncToCSV.py]: responded see comment above about parser definition

File: src/carma-streets/KafkaLogMessage.py (hunk 7 above)
Comment (paulbourelly999) [reply src/carma-streets/KafkaLogMessage.py]: resolved

File: src/carma-streets/KafkaLogMessage.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    SchedulingPlan="scheduling_plan"
+    SDSM="sdsm"
+    DetectedObject="detected_object"
+    VehicleStatusIntent="vehicle_status_intent"
+
+@dataclass
+class KafkaLogMessage:
+    """Class used to store data for each Kafka Log Message
+    """
+    created: int
```
[the remark this answers is not in the document]
Comment (paulbourelly999) [reply src/carma-streets/KafkaLogMessage.py]: Resolved

File: src/carma-streets/ParseKafkaLog.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (adamlm) [reply src/carma-streets/ParseKafkaLog.py]: Makes sense to me

File: src/carma-streets/ParseKafkaLogsDirectory.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        csv_dir (Path): _description_
+    """
+    kafka_log_dir_path = Path(kafka_log_dir)
+    csv_dir_path = Path(csv_dir)
+    if kafka_log_dir_path.is_dir() and not csv_dir_path.is_dir():
+        csv_dir_path.mkdir(exist_ok=False)
+        for kafka_topic_log in kafka_log_dir_path.glob("*.log"):
+            if KafkaLogMessageType.TimeSync.value in kafka_topic_log.name :
+                print(f"Found TimeSync Kafka topic log {kafka_topic_log}. Parsing log to csv ...")
+                ParseTimeSyncToCSV.parse_timesync_to_csv(kafka_topic_log, Path(f"{csv_dir}/{KafkaLogMessageType.TimeSync.value}.csv"))
```
[the remark this answers is not in the document]
Comment (adamlm) [reply src/carma-streets/ParseKafkaLogsDirectory.py]: That's weird. Could it be that you were using `csv_dir` instead of `csv_dir_path`? I will treat this change as optional since it's minor.

**Review by adamlm — APPROVED** [review]
